File: ios/CameraPreviewManager.m
#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>
#import <React/RCTViewManager.h>
#import <React/RCTUtils.h>

@interface RCT_EXTERN_MODULE(CameraPreviewManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(cameraPosition, NSString)
RCT_EXPORT_VIEW_PROPERTY(safeRenderArea, NSDictionary)
RCT_EXPORT_VIEW_PROPERTY(mirrorFramesHorizontally, BOOL)

@end

File: ios/CameraPreviewManager.swift
import Foundation
import SCSDKCameraKit

public class CameraPreviewError: UILabel, CameraViewProtocol {
    @objc public var cameraPosition: NSString = ""

    public init() {
        super.init(frame: CGRect.zero)

        let errorMessage = "Error: [CameraKitReactNative] Attempt to render CameraPreviewView when CameraKit session is not initialized."
        print(errorMessage)

        text = errorMessage
        textColor = .red
        numberOfLines = 0
        lineBreakMode = .byWordWrapping
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

@objc(CameraPreviewManager) public class CameraPreviewManager: RCTViewManager {
    override public static func requiresMainQueueSetup() -> Bool {
        true
    }

    override public func view() -> UIView! {
        guard let cameraKitContext = (bridge.module(for: CameraKitContextModule.self) as? CameraKitContextModule) else {
            return CameraPreviewError()
        }
        return CameraPreviewView(context: cameraKitContext)
    }
}
